### PR TITLE
Refactoring base stuff

### PR DESCRIPTION
**Pull request by jupe**
## Status
**READY**

## Migrations
NO

## Description
Refactoring a bit basic things, like:
* move model registration to own file
* move routes registrations to own file
* move arguments parsing to own file
* move eventBus to own file

### CODE REVIEW

File: app/server.js [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+const sslcert_crt = '../config/sslcert/server.crt';
+
+function createServer(app)
+{
+  if( nconf.get('https') ) {
+      if( !fs.existsSync(sslcert_key) ) {
+          logger.error('ssl cert key is missing: %s', sslcert_key);
+          process.exit(1);
+      }
+      if( !fs.existsSync(sslcert_crt) ) {
```
Comment (olkorhon) [inline app/server.js]: spacing changes in this file, others use eg. 
`if (!fs.existsSync(sslcert_crt))`

File: app/index.js [hunk 2]
```diff
@@ -103,11 +76,18 @@ var onError = function(error){
   }
   process.exit(-1);
 };
-var onListening = function(){
-  var listenurl = (nconf.get('https')?'https':'http:')+'://'+nconf.get('listen')+':'+nconf.get('port');
+function onListening() {
+  let listenurl = (nconf.get('https')?'https':'http:')+'://'+nconf.get('listen')+':'+nconf.get('port');
```
Comment (olkorhon) [inline app/index.js]: could we refactor these nconf.get calls to the top of the file, seems like they are called quite a few times. I assume we are not expecting them to change. Possible names:
* https -> {boolean} const usingHttps
* listen -> const address 
* port -> const port

File: app/index.js [hunk 3]
```diff
@@ -32,69 +39,35 @@ logger.add(require('winston-daily-rotate-file'), {
...
+// create express instance
+const app = Express();
 
-var io = require('socket.io')(server);
+// Create HTTP server.
+const server = Server(app);
 
-// Create public event channel
-global.pubsub = new EventEmitter();
+// register socket io
```
Comment (olkorhon) [inline app/index.js]: majority of the comments start with an uppercase character

File: app/index.js [hunk 4]
```diff
@@ -103,11 +76,18 @@ var onError = function(error){
...
+  let listenurl = (nconf.get('https')?'https':'http:')+'://'+nconf.get('listen')+':'+nconf.get('port');
   console.log('OpenTMI started on ' +listenurl+ ' in '+ nconf.get('cfg')+ ' mode');
+  eventBus.emit('start_listening', {url: listenurl});
 };
 
 server.listen(nconf.get('port'), nconf.get('listen'));
 server.on('error', onError);
 server.on('listening', onListening);
+
+// this would be usefull for testing
```
Comment (olkorhon) [inline app/index.js]: uppercase first letter, also *useful (made that mistake myself a few times.)

File: app/routes/index.js [hunk 5]
```diff
@@ -0,0 +1,25 @@
+// internal modules
+const fs = require('fs');
+
+// 3rd party modules
+const logger = require('winston');
+
+function registerRoutes(app)
+{
```
Comment (olkorhon) [inline app/routes/index.js]: brackets for functions should start from the same line they are introduced
eg. `function foo() {`

File: app/server.js [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+const https = require('https');
+
+// application
+const logger = require('winston');
+const nconf = require('../config');
+const sslcert_key = '../config/sslcert/server.key';
+const sslcert_crt = '../config/sslcert/server.crt';
+
+function createServer(app)
+{
```
Comment (olkorhon) [inline app/server.js]: brackets for functions should start from the same line they are introduced
eg. `function foo() {`

File: app/routes/index.js [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+        !file.match(/error\.js$/)) {
+      logger.verbose(' * '+file);
+      require(`${__dirname}/${file}`)(app);
+    }
+  });
+}
+module.exports.registerRoutes = registerRoutes;
+
+function registerErrorRoute(app)
+{
```
Comment (olkorhon) [inline app/routes/index.js]: brackets for functions should start from the same line they are introduced
eg. `function foo() {`

File: app/models/index.js [hunk 8]
```diff
@@ -0,0 +1,19 @@
+// internal modules
+const fs = require('fs');
+
+// 3rd party modules
+const logger = require('winston');
+
+function registerModels(app)
+{
```
Comment (olkorhon) [inline app/models/index.js]: brackets for functions should start from the same line they are introduced
eg. `function foo() {`

File: app/index.js [hunk 9]
```diff
@@ -103,11 +76,18 @@ var onError = function(error){
...
+module.exports = {
```
Comment (olkorhon) [inline app/index.js]: could use object naming shortcuts and make it a one liner `module.exports = { server, eventBus }`

File: app/index.js [hunk 10]
```diff
@@ -1,18 +1,25 @@
...
+const SocketIO = require('socket.io');
+
+// application modules
+const express = require('./express')
+const Server = require('./server')
 const nconf = require('../config');
+const eventBus = require('./tools/eventBus');
+const models = require('./models');
+const routes = require('./routes');
+const AddonManager = require('./addons');
```
Comment (jupe) [inline app/index.js]: I haven't use `.js`- seems that there is just couple `require()` lines which uses extension `.js`when importing application modules. I propose to use that way for now.

**Comment by jupe** [comment]
seems that db connection parameters causes all this mysterious issues. That need to be investigate in another PR.

File: app/index.js (hunk 10 above)
Comment (olkorhon) [inline app/index.js]: works for me.

File: app/db.js [hunk 11]
```diff
@@ -1,43 +1,59 @@
...
+mongoose.connection.on('error', () => {
     logger.error('Could not connect to MongoDB: ' + dbUrl);
 });
 
-mongoose.connection.on('disconnected', function(){
-    logger.warn('Lost MongoDB connection...');
-    if (!isConnectedBefore)
+mongoose.connection.on('disconnected', () => {
+    logger.warn('MongoDB connection lost');
+    if (!isConnectedBefore) {
```
Comment (olkorhon) [inline app/db.js]: This looks a bit funky, can disconnected really occur before it has connected?

File: app/models/index.js [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+
+
+let models = {};
+
+
+function ensureIndexes() {
+    /** @TODO create mechanism to call this in safe way
+     * so we do not close db connection before the process is completed
+     */
+    logger.info("Start ensuring models indexes..");
```
Comment (olkorhon) [inline app/models/index.js]: we really do need to define those eslint common formatting rules. indent 4 -> 2

Comment (jupe) [inline app/models/index.js]: there is task in backlog - lets pick it in next sprint :)

File: app/controllers/index.js [hunk 13]
```diff
@@ -64,7 +59,9 @@ class DefaultController extends EventEmitter {
   }
 
   find(req, res) {
+    console.log('finding...')
     this._model.query(req.query, (error, list) => {
+      console.log('results...')
```
Comment (jupe) [inline app/controllers/index.js]: no - need to clean

File: app/controllers/index.js [hunk 14]
```diff
@@ -8,34 +8,29 @@ const mongoose = require('mongoose');
...
       logger.debug(`do param ${JSON.stringify(req.params)}`);
       const find = {};
-      find[this.docId] = req.params[modelname];
-
+      find[docId] = req.params[modelname];
       this.Model.findOne(find, (error, data) => {
         if (error) {
-          if (errorCb) errorCb(error);
-          else res.status(300).json({ error });
+          res.status(300).json({ error });
```
Comment (jupe) [inline app/controllers/index.js]: yep wrong statusCode

File: app/db.js (hunk 11 above)
Comment (jupe) [inline app/db.js]: heh yeah :D

File: app/models/index.js (hunk 12 above)
Comment (olkorhon) [inline app/models/index.js]: agreed

**Pull request merged**